strawberryfields/engine.py

```diff
@@ -531,161 +544,507 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+        """
+        return self in (JobStatus.CANCELLED, JobStatus.COMPLETE, JobStatus.FAILED)
+
+
+class Job:
+    """Represents a remote job that can be queried for its status or result.
+
+    This object should typically not be instantiated directly, but returned by an
+    `Engine` or `Connection` when a job is run.
 
     Args:
-        backend (str, BaseBackend): name of the backend, or a pre-constructed backend instance
-        polling_delay_seconds (int): the number of seconds to wait between queries when polling for
-            job results
+        id_ (str): the job ID
+        status (strawberryfields.engine.JobStatus): the job status
+        connection (strawberryfields.engine.Connection): the connection over which the
+            job is managed
     """
 
-    # This engine will execute jobs remotely.
-    REMOTE = True
+    def __init__(self, id_: str, status: JobStatus, connection: "Connection"):
+        self._id = id_
+        self._status = status
+        self._connection = connection
+        self._result = None
 
-    def __init__(self, backend, polling_delay_seconds=1, **kwargs):
-        super().__init__(backend)
+    @property
+    def id(self) -> str:
+        """The job ID.
 
-        api_client_params = {k: v for k, v in kwargs.items() if k in DEFAULT_CONFIG["api"].keys()}
-        self.client = APIClient(**api_client_params)
-        self.polling_delay_seconds = polling_delay_seconds
-        self.jobs = []
+        Returns:
+            str: the job ID
+        """
+        return self._id
 
-    def __str__(self):
-        return self.__class__.__name__ + "({})".format(self.backend_name)
+    @property
+    def status(self) -> JobStatus:
+        """The job status.
+
+        Returns:
+            strawberryfields.engine.JobStatus: the job status
+        """
+        return self._status
 
-    def _init_backend(self, *args):
+    @property
+    def result(self) -> Result:
+        """The job result.
+
+        This is only defined for complete jobs, and raises an exception for any other
+        status.
+
+        Returns:
+            strawberryfields.engine.Result: the job result
         """
-        TODO: This does not do anything right now.
+        if self.status != JobStatus.COMPLETE:
+            raise AttributeError(
+                "The result is undefined for jobs that are not complete "
+                "(current status: {})".format(self.status.value)
+            )
+        return self._result
+
+    def refresh(self):
+        """Refreshes the status of the job, along with the job result if the job is
+        newly completed.
+
+        Only an open or queued job can be refreshed; an exception is raised otherwise.
         """
-        # Do nothing for now...
-        pass
+        if self.status.is_final:
+            raise InvalidJobOperationError(
+                "A {} job cannot be refreshed".format(self.status.value)
+            )
```

Comment:
I'm wondering about raising an error in this case. If the user mistakenly calls on `refresh`, then this would throw an `InvalidJobOperationError` in their script and terminate it. Instead of that, maybe information could be provided, why no refreshing can take place (e.g. the result is already contained, or the job failed, etc.) and perhaps a warning could be raised? :thinking: 

What do you think?

Reply:
(For cancellation, it would seem in order to raise an error, as the user expects to take action, whereas `refresh` seems to be something more along the lines of trying to extract more information about the `Job`. But this might just be my point of view.)

Reply:
Agree

Reply:
Yeah, that makes sense. I'll replace that raised exception with a warning message.

Reply:
That exception has been replaced with a logged warning message.